romancal/datamodels/container.py

```diff
@@ -0,0 +1,486 @@
+        self._models.pop(index)
+
+    def copy(self, memo=None):
+        """
+        Returns a deep copy of the models in this model container.
+        """
+        return copy.deepcopy(self, memo=memo)
+
+    def close(self):
+        if not self._iscopy and self._asdf is not None:
+            self._asdf.close()
+
+    @staticmethod
+    def read_asn(filepath):
+        """
+        Load ASDF files from a Roman association file.
+
+        Parameters
+        ----------
+        filepath : str
+            The path to an association file.
+        """
+        filepath = op.abspath(op.expanduser(op.expandvars(filepath)))
+        try:
+            with open(filepath) as asn_file:
+                asn_data = load_asn(asn_file)
+        except AssociationNotValidError as e:
+            raise OSError("Cannot read ASN file.") from e
+        return asn_data
+
+    def from_asn(self, asn_data, asn_file_path=None):
+        """
+        Load ASDF files from a Roman association file.
+
+        Parameters
+        ----------
+        asn_data : `~roman_datamodels.associations.Association`
+            Association dictionary.
+
+        asn_file_path: str
+            Filepath of the association, if known.
+        """
+        # match the asn_exptypes to the exptype in the association and retain
+        # only those file that match, as a list, if asn_exptypes is set to none
+        # grab all the files
+        if self.asn_exptypes:
+            infiles = []
+            logger.debug(
+                f"Filtering datasets based on allowed exptypes {self.asn_exptypes}:"
+            )
+            for member in asn_data["products"][0]["members"]:
+                if any(
+                    x
+                    for x in self.asn_exptypes
+                    if re.match(member["exptype"], x, re.IGNORECASE)
+                ):
+                    infiles.append(member)
+                    logger.debug(f'Files accepted for processing {member["expname"]}:')
+        else:
+            infiles = list(asn_data["products"][0]["members"])
+
+        asn_dir = op.dirname(asn_file_path) if asn_file_path else ""
+        # Only handle the specified number of members.
+        sublist = infiles[: self.asn_n_members] if self.asn_n_members else infiles
+        try:
+            for member in sublist:
+                filepath = op.join(asn_dir, member["expname"])
+                update_model = any(attr in member for attr in RECOGNIZED_MEMBER_FIELDS)
+                if update_model or self._save_open:
+                    m = rdm.open(filepath, memmap=self._memmap)
+                    m.meta["asn"] = {"exptype": member["exptype"]}
+                    for attr, val in member.items():
+                        if attr in RECOGNIZED_MEMBER_FIELDS:
+                            if attr == "tweakreg_catalog":
+                                val = op.join(asn_dir, val) if val.strip() else None
+                            setattr(m.meta, attr, val)
```

Comment:
A test covering lines 317-320 i needed. Isn't this what tweakreg is going to use? Let's make sure it works.